nav2_costmap_2d/src/costmap_cleaner.cpp

```diff
@@ -0,0 +1,145 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <vector>
+#include <string>
+#include <algorithm>
+#include <memory>
+
+#include "nav2_costmap_2d/costmap_cleaner.hpp"
+#include "nav2_costmap_2d/costmap_2d_ros.hpp"
+
+namespace nav2_costmap_2d
+{
+
+using std::vector;
+using std::string;
+using std::shared_ptr;
+using std::any_of;
+using CleanCostmap = nav2_msgs::srv::CleanCostmap;
+
+CostmapCleaner::CostmapCleaner(rclcpp::Node::SharedPtr & node, Costmap2DROS * costmap)
+: node_(node), costmap_(costmap)
+{
+  node_->get_parameter_or_set("cleanable_layers", cleanable_layers_, {"obstacle_layer"});
+
+  server_ = node_->create_service<CleanCostmap>("clean_" + costmap->getName(),
+      std::bind(&CostmapCleaner::cleanCallback, this,
+      std::placeholders::_1, std::placeholders::_2, std::placeholders::_3));
+}
+
+void CostmapCleaner::cleanCallback(
+  const shared_ptr<rmw_request_id_t>/*request_header*/,
+  const shared_ptr<CleanCostmap::Request> request,
+  const shared_ptr<CleanCostmap::Response>/*response*/)
+{
+  RCLCPP_INFO(node_->get_logger(), "Received request to clean " + costmap_->getName());
+
+  if (costmap_ == nullptr) {
+    RCLCPP_ERROR(node_->get_logger(), "Costmap is undefined. Doing nothing.");
+    return;
+  }
+
+  RCLCPP_INFO(node_->get_logger(), "Will proceed with cleaning the costmap");
+
+  clean(request->reset_distance);
+}
+
+void CostmapCleaner::clean(const double reset_distance)
+{
+  double x, y;
+
+  if (!getPose(x, y)) {
+    RCLCPP_ERROR(node_->get_logger(), "Cannot clean map because robot pose cannot be retrieved.");
+    return;
+  }
+
+  auto layers = costmap_->getLayeredCostmap()->getPlugins();
+
+  for (auto & layer : *layers) {
+    auto name = getLayerName(*layer);
+
+    if (any_of(
+        begin(cleanable_layers_), end(cleanable_layers_), [&name](auto l) {return l == name;}))
+    {
+      auto costmap_layer = std::static_pointer_cast<CostmapLayer>(layer);
+      cleanLayer(costmap_layer, x, y, reset_distance);
+    }
+  }
+}
+
+void CostmapCleaner::cleanLayer(
+  shared_ptr<CostmapLayer> & costmap, double pose_x, double pose_y, double reset_distance)
+{
+  std::unique_lock<Costmap2D::mutex_t> lock(*(costmap->getMutex()));
+
+  double start_point_x = pose_x - reset_distance / 2;
+  double start_point_y = pose_y - reset_distance / 2;
+  double end_point_x = start_point_x + reset_distance;
+  double end_point_y = start_point_y + reset_distance;
+
+  int start_x, start_y, end_x, end_y;
+  costmap->worldToMapNoBounds(start_point_x, start_point_y, start_x, start_y);
+  costmap->worldToMapNoBounds(end_point_x, end_point_y, end_x, end_y);
+
+  unsigned char * grid = costmap->getCharMap();
+
+  for (int x = 0; x < static_cast<int>(costmap->getSizeInCellsX()); x++) {
+    bool isOutXrange = x<start_x || x> end_x;
```

Comment:
You should first find the range of indices and only iterate over those, maps can be massive (think far larger than 100x100 meters), iterating over the full window for a subsection would be very costly

Reply:
@orduno Why not stick with the function that already exists `setConvexPolygonCost`

Reply:
I didn't provide enough info on the PR description and that caused some confusion. -- We are clearing everything except some region around the robot. 

I updated the code to use a more efficient method for clearing.